server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1547,71 +1551,71 @@ public void regenerateCertificatesOf(Owner owner, String productId, boolean lazy
-            deletePool(dp);
         }
 
-        pool.getEntitlements().remove(entitlement);
-        poolCurator.merge(pool);
-        entitlementCurator.delete(entitlement);
+        poolCurator.lock(poolsToLock);
+
+        log.info("Batch revoking {} entitlements ", entsToRevoke.size());
 
-        Event event = eventFactory.entitlementDeleted(entitlement);
-        if (!entitlement.isValid() &&
-                entitlement.getPool().isUnmappedGuestPool() &&
-                consumerCurator.getHost(consumer.getFact("virt.uuid"), consumer.getOwner()) == null) {
-            event = eventFactory.entitlementExpired(entitlement);
-            event.setMessageText(event.getMessageText() + ": " +
-                i18n.tr("Unmapped guest entitlement expired without establishing a host/guest mapping."));
+        entsToRevoke =  new ArrayList<Entitlement>(entsToRevoke);
+
+
+        for (Pool pool : poolsToDelete) {
+            entsToRevoke.addAll(pool.getEntitlements());
         }
 
-        // The quantity is calculated at fetch time. We update it here
-        // To reflect what we just removed from the db.
-        pool.setConsumed(pool.getConsumed() - entitlement.getQuantity());
+        poolCurator.lock(poolsToDelete);
 
-        if (consumer.getType().isManifest()) {
-            pool.setExported(pool.getExported() - entitlement.getQuantity());
+        for (Entitlement ent : entsToRevoke) {
+            //We need to trigger lazy load of provided products
+            //to have access to those products later in this method.
+            ent.getPool().getProvidedProducts().size();
+            Pool pool = ent.getPool();
+            pool.setConsumed(pool.getConsumed() - ent.getQuantity());
+
+            Consumer consumer = ent.getConsumer();
+            if (consumer.getType().isManifest()) {
+                pool.setExported(pool.getExported() - ent.getQuantity());
+            }
         }
 
-        // Check for a single stacked sub pool as well. We'll need to either
-        // update or delete the sub pool now that all other pools have been deleted.
-        if (!"true".equals(pool.getAttributeValue("pool_derived")) &&
-            pool.getProduct().hasAttribute("stacking_id")) {
+        log.info("Starting batch delete of pools and entitlements");
+        poolCurator.batchDelete(poolsToDelete);
+        entitlementCurator.batchDelete(entsToRevoke);
+        entitlementCurator.flush();
+        log.info("All deletes flushed successfully");
+
+        List<Entitlement> stackingEntitlements = filterStackingEntitlements(entsToRevoke);
 
-            String stackId = pool.getProduct().getAttributeValue("stacking_id");
+        for (Entitlement ent : stackingEntitlements) {
+            Pool pool = ent.getPool();
+            Consumer consumer = ent.getConsumer();
+            String stackId = pool.getProductAttributeValue("stacking_id");
             Pool stackedSubPool = poolCurator.getSubPoolForStackId(consumer, stackId);
```

Comment:
consider using the [batch impl](https://github.com/candlepin/candlepin/pull/1073/files#diff-0cac811ddb60a663206629eab27f0c12R641) of this from #1073